Custom reports updates

carkom opened the pull request:
Some slight UI and visual changes to custom reports

Teprifer commented:
Confirming categories sort per budget order when importing my budget.

Teprifer commented:
Ok this is interesting, thought I'd dig a little deeper as I originally just checked my regular deposits filter in the default and came across some odd behaviour now:

1. Click create new custom report
2. Click **Type** and _Net_ is disabled
3. Change **Split By** to _Account_
4. Change **Split By** back to _Category_
5. Click **Type** and _Net_ is now available for selection

![image](https://github.com/actualbudget/actual/assets/34882928/821777f5-71af-471b-a4e4-efa85ec10ae5)
_By way of explanation since I cut off categories, the far right is "Income" and the one in the middle is a category with Deposit and Payments categorised to it, net positive_


For comparison, this is the same graph, just changed **Type** back to _Payment_

![image](https://github.com/actualbudget/actual/assets/34882928/e2eb6c08-57a5-4b76-8dd2-a09c4e0efad1)

youngcw commented:
This is nice so far.  I think it would be better long term to try to not use so much screen space for the create new report button.  I would be a fan of a floating button, maybe at the bottom corner.

MatissJanis commented:
> This is nice so far. I think it would be better long term to try to not use so much screen space for the create new report button. I would be a fan of a floating button, maybe at the bottom corner.

+1

IMO a better location for the button would be at the top of the page (screenshot). Plus we should make it smaller and `primary` color.

And then we can get rid of the separator line between the graphs. 

<img width="1440" alt="Screenshot 2024-02-28 at 08 01 57" src="https://github.com/actualbudget/actual/assets/886567/aa0137b9-1abd-40b7-8bbc-215ad4748aee">

youngcw commented:
sankey report now takes up a whole row in the dash

carkom commented:
> After this is there anything else you think NEEDS to be included? or is this about ready to remove the feature flag?

There's two things I'd like to finish before removing feature flag. 
1. incorporating "yearly" interval choice (I have a PR about 75% done)
2. create "disabledList" to make the functionality more consistent and the code easier to read (This PR is ready to submit)

My assumption is that the next release should have these PRs merged and a further PR to remove feature flag. Cheers!

youngcw commented:
vrt needs fixed

carkom commented:
> vrt needs fixed

Was just an errored out test, no VRT changes. Cheers!

MatissJanis reviewed (changes requested):
The old widget size was visually much more pleasant. Can we revert this change back? There should not be any VRT screenshot updates in this PR. It should only update things to do with custom reports.

MatissJanis commented:
> You mean the networth and cash flow cards?

Yes

MatissJanis commented:
<img width="624" alt="Screenshot 2024-02-29 at 19 41 12" src="https://github.com/actualbudget/actual/assets/886567/45970571-49de-484d-a894-1a7c451ff5ba">

Found one more bug here: the spacing between the header and the other content is too small.

carkom commented:
Both fixed. Cheers!

MatissJanis commented on packages/desktop-client/src/components/reports/reports/CustomReportListCards.tsx:
```diff
@@ -61,20 +63,57 @@ export function CustomReportListCards({
+
+    if (response.error) {
+      setErr(response.error.message);
+      onNameMenuOpen(reportData.id === undefined ? '' : reportData.id, true);
+      return;
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏we can remove these lines to reduce the amount of duplication
```suggestion
```

carkom replied:
This is not dupliacted. The state "true" keeps the menu open so the user can read the error. With this removed the error gets written to a menu that then immediately gets closed without time for the user to read it.

MatissJanis reviewed (dismissed):
Just a small nitpick on my end. Overall: LGTM!

(but please wait for another approval before merging as quite a few others were looking into this PR too)

twk3 commented on packages/desktop-client/src/components/reports/Overview.jsx:
```diff
@@ -29,6 +28,20 @@ export function Overview() {
+            marginTop: 10,
+          }}
+        >
+          <ButtonLink to="/reports/custom" type="primary">
```
This is currently navigating to the last loaded custom report rather than to a new custom report if used after looking at a saved report. I think we want it to open a default report each time?

carkom replied:
I'm not experiencing the same behavior. If I open a saved report then click "reports" in the main menu sidebar then click this button, I get a new default report. Is your experience different? How can I duplicate what you're seeing?

twk3 replied:
Looks like it starts happening if you load a custom report. Then go back to the cards and rename it from there. Then after the rename, click the button. (and it loads with the old name, prior to the rename)

carkom replied:
I think I got it sorted, can you test again?

twk3 commented on packages/desktop-client/src/components/reports/ReportOptions.ts:
```diff
@@ -150,7 +150,15 @@ export const categoryLists = (categories: {
+    ...categories.list.sort((a, b) => {
+      const catGroupA = categories.grouped.find(f => f.id === a.cat_group);
+      const catGroupB = categories.grouped.find(f => f.id === b.cat_group);
+      return a.sort_order && b.sort_order && catGroupA && catGroupB
```
suggestion: Can we comment the return to better describe the condition going on here. It's a pretty intense condition to just read later without context as to why.

twk3 commented on packages/desktop-client/src/components/reports/reports/CustomReport.jsx:
```diff
@@ -312,7 +312,23 @@ export function CustomReport() {
+            color: theme.pageTextPositive,
+          }}
+        >
+          {report.name ?? 'Unsaved report'}
```
```suggestion
          {report.name || 'Unsaved report'}
```

Nullish operator isn't going to work here, because we have an empty string (falsey) rather than nullish.

Teprifer commented:
Appreciate adding the saved report name to the header, 2 verrry minor suggestions 'for another day' :) 

1) Remove the 's' in reports to make it singular as it's a single report being viewed.
2) Add the (modified) tag when doing so with right save/load menu.

e.g.

![image](https://github.com/actualbudget/actual/assets/34882928/85a7f187-04e5-4054-af8b-0341e75aa372)

carkom commented:
> Appreciate adding the saved report name to the header, 2 verrry minor suggestions 'for another day' :)
> 
>     1. Remove the 's' in reports to make it singular as it's a single report being viewed.
> 
>     2. Add the (modified) tag when doing so with right save/load menu.

I removed the "s" as that's an easy one. I've added the second suggestion to my feedback doc for a future PR.

twk3 reviewed (approved):
The changes look good. Thanks @carkom
